solidity/contracts/contracts/QuantumGravityBridge.sol

```diff
@@ -0,0 +1,314 @@
+    /// A hash of all relevant information about a message root.
+    /// The format of the hash is:
+    ///     keccak256(bridge_id, bytes32("transactionBatch"), nonce, message_root)
+    /// @param _bridge_id Bridge ID.
+    /// @param _nonce Nonce.
+    /// @param _messageRoot Message root.
+    function domainSeparateMessageRoot(
+        bytes32 _bridge_id,
+        uint256 _nonce,
+        bytes32 _messageRoot
+    ) private pure returns (bytes32) {
+        // bytes32 encoding of the string "transactionBatch"
+        bytes32 methodName = 0x7472616e73616374696f6e426174636800000000000000000000000000000000;
+
+        bytes32 c = keccak256(abi.encode(_bridge_id, methodName, _nonce, _messageRoot));
+
+        return c;
+    }
+
+    /// @dev Checks that enough voting power signed over a digest.
+    /// @param _currentValidators The current validators.
+    /// @param _sigs The current validators' signatures.
+    /// @param _digest This is what we are checking they have signed.
+    /// @param _powerThreshold At least this much power must have signed.
+    function checkValidatorSignatures(
+        // The current validator set and their powers
+        Validator[] calldata _currentValidators,
+        Signature[] calldata _sigs,
+        bytes32 _digest,
+        uint256 _powerThreshold
+    ) private pure {
+        uint256 cumulativePower = 0;
+
+        for (uint256 i = 0; i < _currentValidators.length; i++) {
+            // Check that the current validator has signed off on the hash.
+            if (!verifySig(_currentValidators[i].addr, _digest, _sigs[i])) {
+                revert InvalidSignature();
+            }
+
+            // Sum up cumulative power.
+            cumulativePower += _currentValidators[i].power;
+
+            // Break early to avoid wasting gas.
+            if (cumulativePower >= _powerThreshold) {
+                break;
+            }
+        }
+
+        // Check that there was enough power
+        if (cumulativePower < _powerThreshold) {
+            revert InsufficientVotingPower();
+        }
+    }
+
+    /// @notice This updates the validator set by checking that the validators
+    /// in the current validator set have signed off on the new validator set.
+    /// The signatures supplied are the signatures of the current validator set
+    /// over the checkpoint hash generated from the new validator set. Anyone
+    /// can call this function, but they must supply valid signatures of the
+    /// current validator set over the new validator set.
+    /// @param _newValidatorSetHash The hash of the new validator set
+    /// @param _newNonce The new nonce.
+    /// @param _currentValidatorSet The current validator set.
+    /// @param _sigs Signatures.
+    function updateValidatorSet(
+        uint256 _newNonce,
+        uint256 _newPowerThreshold,
+        bytes32 _newValidatorSetHash,
+        Validator[] calldata _currentValidatorSet,
```

Comment:
Is this the new validator set?

Reply:
No, this is the current validator set! The new validator set is committed to in `_newValidatorSetHash`.